File: src/com/dixie/adventofcode/lib/Day.java
package com.dixie.adventofcode.lib;

import java.util.List;
import java.util.Scanner;
import java.util.stream.Collectors;

public abstract class Day {

  public void solve() {
    System.out.println("Input:");
    Scanner scan = new Scanner(System.in);
    List<String> lines = StreamUtils.streamLines(scan)
            .takeWhile(line -> !line.equals("end"))
            .collect(Collectors.toList());
    System.out.println("Part 1: " + solve(lines, true));
    System.out.println("Part 2: " + solve(lines, false));
  }

  protected Object solve(List<String> lines, boolean part1) {
    return part1 ? part1(lines) : part2(lines);
  }

  protected Object part1(List<String> lines) {
    return 0;
  }

  protected Object part2(List<String> lines) {
    return 0;
  }
}

File: src/com/dixie/adventofcode/aoc2022/lib/Day.kt
package com.dixie.adventofcode.aoc2022.lib

abstract class Day {
  protected lateinit var lines: List<String>

  fun run() {
    println("Input:")
    val lines = generateSequence { readlnOrNull() }.takeWhile { it != null && it != "end" }.toList()
    prepare()
    println("Part 1: ${solve(true)}")
    println("Part 2: ${solve(false)}")
  }

  abstract fun prepare()

  open fun solve(part1: Boolean) = if (part1) part1() else part2()

  abstract fun part1(): String

  abstract fun part2(): String
}

fun String.toInts(delimiter: String = " ") = splitToSequence(delimiter).map(String::toInt)
fun String.toLongs(delimiter: String = " ") = splitToSequence(delimiter).map(String::toLong)
fun String.toDoubles(delimiter: String = " ") = splitToSequence(delimiter).map(String::toDouble)